File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.java
class Solution {
    
    // TrieNode
    static class Node {
        Map<Character, Node> children;
        Map<Character, Integer> count;
        Node() {
            children = new HashMap<>();
            count = new HashMap<>();
        }
        boolean contains(char c) {
            return count.containsKey(c);
        }
    }
    
    public int[] sumPrefixScores(String[] words) {
        Node root = new Node();
        // add words to Trie and update counts
        for(String word: words) {
            Node curr = root;
            for (char c: word.toCharArray()) {
                if (!curr.contains(c)) {
                    curr.children.put(c, new Node());
                    curr.count.put(c, 0);
                }
                curr.count.put(c, curr.count.get(c) + 1);
                curr = curr.children.get(c);
            }
        }
        int[] ret = new int[words.length];
        // find answer for the ith word
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            int cnt = 0;
            Node curr = root;
            for (char c: word.toCharArray()) {
                if (curr.contains(c)) {
                    cnt += curr.count.get(c);
                    curr = curr.children.get(c);
                }
                else {
                    break;
                }
            }
            ret[i] = cnt;
        }
        return ret;
    }
}

File: 2416-sum-of-prefix-scores-of-strings/2416-sum-of-prefix-scores-of-strings.py
class Node:
    def __init__(self):
        self.children = [None] * 26
        self.cnt = 0
        
class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        root = Node()
        for word in words:
            curr = root
            for i in word:
                ind = ord(i) - ord('a')
                if curr.children[ind] is None:
                    curr.children[ind] = Node()
                curr = curr.children[ind]
                curr.cnt += 1
        
        ret = []
        for word in words:
            curr = root
            nxt = 0
            for i in word:
                ind = ord(i) - ord('a')
                if curr.children[ind] is not None:
                    curr = curr.children[ind]
                    nxt += curr.cnt
                else:
                    break
            ret.append(nxt)
        return ret